### PR TITLE
[refactoring] General maintenance for code structure

### LINKED ISSUE #903: [code-structure] Add all STM32 Core-IDs to the codebase for MCU identification
We have Core-IDs defined in `/include/stlink/stm32.h` which are used to detect chips correctly.
However this is implemented differently for the various STM32 product lines which is not very clever nor follows any visible logic structure. The current situation is as follows:

```
/* STM32 Cortex-M core ids (CPUTAPID) */

#define STM32_CORE_ID_M0_SWD 0x0bb11477
// (RM0091 Section 32.5.3) F0 SW-DP
// (RM0444 Section 40.5.3) G0 SW-DP

#define STM32_CORE_ID_M0P_SWD 0x0bc11477
// (RM0385 Section 27.5.3) L0 SW-DP

#define STM32_CORE_ID_M3_r1p1_SWD 0x1ba01477
// (RM0008 Section 31.8.3) F1 SW-DP
#define STM32_CORE_ID_M3_r1p1_JTAG 0x3ba00477
// (RM0008 Section 31.6.3) F1 JTAG

#define STM32_CORE_ID_M3_r2p0_SWD 0x2ba01477
// (RM0033 Section 32.8.3) F2 SW-DP
// (RM0038 Section 30.8.3) L1 SW-DP
#define STM32_CORE_ID_M3_r2p0_JTAG 0x0ba00477
// (RM0033 Section 32.6.3) F2 JTAG
// (RM0038 Section 30.6.2) L1 JTAG

#define STM32_CORE_ID_M4_r0p1_SWD 0x1ba01477
// (RM0316 Section 33.8.3) F3 SW-DP
// (RM0351 Section 48.8.3) L4 SW-DP
// (RM0432 Section 57.8.3) L4+ SW-DP
#define STM32_CORE_ID_M4_r0p1_JTAG 0x4ba00477
// (RM0316 Section 33.6.3) F3 JTAG
// (RM0351 Section 48.6.3) L4 JTAG
// (RM0432 Section 57.6.3) L4+ JTAG

#define STM32_CORE_ID_M4F_r0p1_SWD 0x2ba01477
// (RM0090 Section 38.8.3) F4 SW-DP
// (RM0090 Section 47.8.3) G4 SW-DP
#define STM32_CORE_ID_M4F_r0p1_JTAG 0x4ba00477
// (RM0090 Section 38.6.3) F4 JTAG
// (RM0090 Section 47.6.3) G4 JTAG

#define STM32_CORE_ID_M7F_SWD 0x5ba02477
// (RM0385 Section 40.8.3) F7 SW-DP
#define STM32_CORE_ID_M7F_JTAG 0x5ba00477
// (RM0385 Section 40.6.3) F7 JTAG

#define STM32_CORE_ID_M7F_H7_SWD 0x6ba02477
// (RM0433 Section 60.4.1) H7 SW-DP
#define STM32_CORE_ID_M7F_H7_JTAG 0x6ba00477
// (RM0433 Section 60.4.1) H7 JTAG

#define STM32_CORE_ID_M33_SWD 0x0be02477
// (RM0438 Section 52.2.10) L5 SW-DP
// (RM0456 Section 65.3.3) U5 SW-DP
#define STM32_CORE_ID_M33_JTAGD 0x0be01477
// (RM0438 Section 52.2.10) L5 JTAG-DP
// (RM0456 Section 65.3.3) U5 JTAG-DP
#define STM32_CORE_ID_M33_JTAG 0x0ba04477
// (RM0438 Section 52.2.8) L5 JTAG
// (RM0456 Section 56.3.1) U5 JTAG
```

In reference to #756 and #761 this seems not to have been implemented well over time and should be commonly addressed.

### LINKED ISSUE #1090: [feature] STM32WLx5 Support
Currently the WLx5 family are not supported. 
I would like to open a PR to merge WLx5 support based on the work done for the WB family of chips.
The only real issue is that the WL55 and H7 have the same core_id.

- [x] Programmer/board type: all
- [x] Operating system: all
- [x] Stlink tools version and/or git commit hash: develop/git-4bfaab0
- [x] Stlink commandline tool name: st-info`, `st-flash`, `st-util`
- [x] Target chip (and optional board): STM32WLE5

### LINKED ISSUE #1199: Wrong flash size reported for various chips
I'm debugging firmware on STM32H725VGT that has some code above 0x8020000. When I try to access this memory e.g. via `disas` command in GDB, I get _"Cannot access memory at address 0x802xxxx"_. I've tracked it down to `info mem` reporting flash region from 0x08000000 to 0x08011c00. The 0x11c00 size is also what I get from `st-info --flash`. This size is wrong: the chip has 1MiB of flash.

I've also tried this on another chip, the one on STM32F407G-DISC1, and `st-info --flash` reports flash size of 0x3fffc00, which would be almost 64MiB! The actual size is still 1MiB.

I'm running stlink from `develop` branch, at e97f438075ff5128630dfa3668f6ac7bd1181c51. OS is Ubuntu 20.04 (Linux 5.4.0-65-generic).

### LINKED ISSUE #1212: [refactoring] General maintenance
The following tasks should be achieved to improve the codebase of the toolset:

- modularise the codebase
- reduce cross-dependencies between headers and source files where possible
- separate device-specific settings from source code (#1129)
- rewrite and modularisation of: `chipid.c` `chipid.h` `common.c` `stlink.h` `stm32.h`
- allow general use of core-IDs for MCU identification (#903)

Some intentions on general proceedings (priorised):
1) Simplification / Clean-up / Restructuring
2) Correction of data types based on C/C++ Reference and common recommendations
3) Evaluation on how to commonly include Cortex-M core ids (CPUTAPID) in the workflow of device identification
4) Definition of structures to prepare for future ABI compatibilty
5) Update for the `stlink-lib` documentation

### PR DESCRIPTION
**Pull request by Nightwalker-87**
List of changes and improvements:

- Revised device-specific parametres in the .chip files
- Aligned naming for .chip files and defines
- Unified formatting for .chip files
- Removed function `dump_a_chip` and old device database from `common.c`
- Moved-out STM32-specific defines to `stm32.h`
- Updated MCU core-ids
- Removed additional `extern "C"` linkage specs

(Closes #903) (Closes #1090) (Closes #1199) (Closes #1212)

### CODE REVIEW

File: src/win32/mmap.h [hunk 1]
```diff
@@ -13,16 +13,10 @@
 #define MAP_ANONYMOUS (1 << 5)
 #define MAP_FAILED    ((void *)-1)
 
-#ifdef __cplusplus
-extern "C" {
-#endif
```
Comment (gszy) [inline src/win32/mmap.h:16]:
```suggestion
```

File: src/win32/mmap.h [hunk 2]
```diff
@@ -13,16 +13,10 @@
...
 
 void *mmap(void *addr, size_t len, int prot, int flags, int fd, long long offset);
 int munmap(void *addr, size_t len);
 
-#ifdef __cplusplus
-}
-#endif
```
Comment (gszy) [inline src/win32/mmap.h:20]:
```suggestion
```

File: src/win32/sys_time.h [hunk 3]
```diff
@@ -12,15 +12,8 @@ struct timezone {
...
 
-#ifdef __cplusplus
-extern "C" {
-#endif
-
 int gettimeofday(struct timeval *tv, struct timezone *tz);
 
-#ifdef __cplusplus
-}
-#endif
```
Comment (gszy) [inline src/win32/sys_time.h:17]:
```suggestion
```

File: src/stlink-lib/usb.h [hunk 4]
```diff
@@ -73,8 +70,5 @@ stlink_t *stlink_open_usb(enum ugly_loglevel verbose, enum connect_type connect,
 size_t stlink_probe_usb(stlink_t **stdevs[], enum connect_type connect, int freq);
 void stlink_probe_usb_free(stlink_t **stdevs[], size_t size);
 
-#ifdef __cplusplus
-}
-#endif
```
Comment (gszy) [inline src/stlink-lib/usb.h:73]:
```suggestion
```

File: src/stlink-lib/usb.h [hunk 5]
```diff
@@ -12,9 +12,6 @@
 #include <libusb_settings.h>
 #include "logging.h"
 
-#ifdef __cplusplus
-extern "C" {
-#endif
```
Comment (gszy) [inline src/stlink-lib/usb.h:15]:
```suggestion
```

File: src/stlink-lib/sg.h [hunk 6]
```diff
@@ -60,8 +57,5 @@ struct stlink_libsg {
 
 stlink_t* stlink_v1_open(const int verbose, int reset);
 
-#ifdef __cplusplus
-}
-#endif
```
Comment (gszy) [inline src/stlink-lib/sg.h:60]:
```suggestion
```

File: src/stlink-lib/sg.h [hunk 7]
```diff
@@ -9,9 +9,6 @@
 #include <stlink.h>
 #include <libusb_settings.h>
 
-#ifdef __cplusplus
-extern "C" {
-#endif
```
Comment (gszy) [inline src/stlink-lib/sg.h:12]:
```suggestion
```

File: src/stlink-lib/flash_loader.h [hunk 8]
```diff
@@ -12,16 +12,10 @@
...
-extern "C" {
-#endif
 
 int stlink_flash_loader_init(stlink_t *sl, flash_loader_t* fl);
 int stlink_flash_loader_write_to_sram(stlink_t *sl, stm32_addr_t* addr, size_t* size);
 int stlink_flash_loader_run(stlink_t *sl, flash_loader_t* fl, stm32_addr_t target, const uint8_t* buf, size_t size);
 
-#ifdef __cplusplus
-}
-#endif
```
Comment (gszy) [inline src/stlink-lib/flash_loader.h:20]:
```suggestion
```

File: src/stlink-lib/flash_loader.h [hunk 9]
```diff
@@ -12,16 +12,10 @@
 
 #include <stlink.h>
 
-#ifdef __cplusplus
-extern "C" {
-#endif
```
Comment (gszy) [inline src/stlink-lib/flash_loader.h:15]:
```suggestion
```

**Review by gszy — APPROVED** [review]
Looks good, though I don’t have the time to review it really carefully.

**Comment by Nightwalker-87** [comment]
@gszy I'll clean-up the white space stuff in a separate commit.

File: src/win32/mmap.h (hunk 1 above)
Comment (Nightwalker-87) [reply src/win32/mmap.h]: Addressed in commit da3d9e3f1e1a6d116df5830cf506f30ff56196b3.

File: src/win32/sys_time.h (hunk 3 above)
Comment (Nightwalker-87) [reply src/win32/sys_time.h]: Addressed in commit da3d9e3f1e1a6d116df5830cf506f30ff56196b3.

File: src/stlink-lib/usb.h (hunk 4 above)
Comment (Nightwalker-87) [reply src/stlink-lib/usb.h]: Addressed in commit da3d9e3f1e1a6d116df5830cf506f30ff56196b3.

File: src/stlink-lib/sg.h (hunk 6 above)
Comment (Nightwalker-87) [reply src/stlink-lib/sg.h]: Addressed in commit da3d9e3f1e1a6d116df5830cf506f30ff56196b3.

File: src/stlink-lib/sg.h (hunk 7 above)
Comment (Nightwalker-87) [reply src/stlink-lib/sg.h]: Addressed in commit da3d9e3f1e1a6d116df5830cf506f30ff56196b3.

File: src/stlink-lib/flash_loader.h (hunk 8 above)
Comment (Nightwalker-87) [reply src/stlink-lib/flash_loader.h]: Addressed in commit da3d9e3f1e1a6d116df5830cf506f30ff56196b3.

File: src/stlink-lib/flash_loader.h (hunk 9 above)
Comment (Nightwalker-87) [reply src/stlink-lib/flash_loader.h]: Addressed in commit da3d9e3f1e1a6d116df5830cf506f30ff56196b3.

**Comment by Nightwalker-87** [comment]
@Ant-ON Could you review this in order to move on?

**Comment by Nightwalker-87** [comment]
This PR will be merged by Friday. Please consider this if you intend to review.

**Pull request merged**